### PR TITLE
Add workhorse TaskPool.

### LINKED ISSUE #1919: Performance regression on 3.1 with mongo backend/maxtasksperchild=1
Demo app: https://gist.github.com/ionelmc/9517684
## With 3.0

Run `./bootstrap.sh "<3.1" mongodb` and:

```
$ .ve/bin/python app.py worker

 -------------- celery@minibox v3.0.24 (Chiastic Slide)
---- **** -----
--- * ***  * -- Linux-3.11.0-17-generic-x86_64-with-Ubuntu-12.04-precise
-- * - **** ---
- ** ---------- [config]
- ** ---------- .> broker:      amqp://test@localhost:5672/test
- ** ---------- .> app:         __main__:0x15fa990
- ** ---------- .> concurrency: 7 (processes)
- *** --- * --- .> events:      OFF (enable -E to monitor this worker)
-- ******* ----
--- ***** ----- [queues]
 -------------- .> celery:      exchange:celery(direct) binding:celery
                .> minibox.dq:  exchange:C.dq(direct) binding:minibox

[pid=39695 processName=MainProcess - 2014-03-13 00:08:08,646]: celery.redirected - WARNING - celery@minibox ready.
```

and:

```
$ .ve/bin/python app.py test 10000
532.74 tasks/sec
$ .ve/bin/python app.py test 10000   
574.35 tasks/sec
$ .ve/bin/python app.py test 10000   
570.28 tasks/sec
```
## With 3.1

Run `./bootstrap.sh "[mongodb]<3.2"` and:

```
$ .ve/bin/python app.py worker

 -------------- celery@minibox v3.1.9 (Cipater)
---- **** -----
--- * ***  * -- Linux-3.11.0-17-generic-x86_64-with-Ubuntu-12.04-precise
-- * - **** ---
- ** ---------- [config]
- ** ---------- .> app:         __main__:0x1299450
- ** ---------- .> transport:   amqp://test@localhost:5672/test
- ** ---------- .> results:     mongodb://localhost/test
- *** --- * --- .> concurrency: 7 (prefork)
-- ******* ----
--- ***** ----- [queues]
 -------------- .> celery           exchange=celery(direct) key=celery
                .> celery@minibox.dq exchange=C.dq(direct) key=celery@minibox

[pid=57621 processName=MainProcess - 2014-03-13 00:22:31,129]: celery.redirected - WARNING - celery@minibox ready.
```

and:

```
$ .ve/bin/python app.py test 1000
59.58 tasks/sec
$ .ve/bin/python app.py test 1000 
60.88 tasks/sec
```

### PR DESCRIPTION
**Pull request by ionelmc**
Fixes #1919

### CODE REVIEW

**Comment by thedrow** [comment]
The idea itself is awesome but this has a few problems that might prevent it being merged:
1) The usage of ctypes might be a problem on PyPy. Is there an alternative?
2) Some of the commits should be squashed and the entire branch should be rebased. I guess I can do this myself if required.
3) This requires more tests.
4) No docstrings.
5) The documentation doesn't specify how many tasks per child is sane.

**Comment by ionelmc** [comment]
Hmmm, this has been on my todo list for quite a while now.

Regarding your points:
1) It's fine on PyPy - just the stucture class is used. Ctypes is not used for calling any APIs
3) Yes indeed. There are also some features that this worker is missing (but can be added):

* IIRC the shutdown needs some improvements (the cold/warm shutdowns)
* No support for remote commands (eg: changing the concurrency)

5) It's the same as for the prefork pool. You should prepare for the worst case as usual :-)

Do you want to help with (3)?

**Comment by thedrow** [comment]
3) I can certainly try to help.
5) I'm referring to the documentation you added here: https://github.com/celery/celery/pull/1928/files#diff-b8704ee1fdde7e4adc4306fa85614015R262
Does this pool fit only when you need to isolate your processes or is a better fit on Linux in general?

Care to rebase this PR so we'll check that nothing is broken.

File: celery/concurrency/workhorse.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        self.sigfh.close()
+        signalfd.sigprocmask(signalfd.SIG_BLOCK, [])
+
+    def on_sigchld(self):
+        pending = {}
+
+        si = signalfd_siginfo()
+        while True:
+            try:
+                if not self.sigfh.readinto(si):
```
Comment (thedrow) [inline celery/concurrency/workhorse.py]: I understand why you are doing this but I'm really not 100% sure that is will be performant when using PyPy.

**Comment by thedrow** [comment]
Also see https://github.com/larsks/python-signalfd/issues/1
We probably need to get this done before merging this.

**Comment by ionelmc** [comment]
@thedrow Regarding signalfd, this is gonna do the trick right? http://blitz.works/butter/file/63d6906419ec/butter/_signalfd.py

It's released on PyPI https://pypi.python.org/pypi/butter/0.11.1

**Comment by thedrow** [comment]
Looks Ok but what's the difference between it and python-signalfd?

**Comment by ionelmc** [comment]
It's not GPL for starters :)

**Comment by thedrow** [comment]
Yup, that's a good reason :)

**Comment by ionelmc** [comment]
Does @ask have any opinion if this should be a separate library on pypi?

**Comment by ask** [comment]
The big problem with this is that it only works on Linux, if it had the ability to work on BSD it would fit better in core.  The C extension requirement is also quite annoying, maybe it's possible to vendorize that in billiard (not sure what the license is)

**Comment by ionelmc** [comment]
I could also do this as an "addon" package (like `celery-workhorse` on pypi) - the worker pool is 
swapable.

It's certainly possible to avoid signalfd - I just need more time to consider/add a signafd-less mode.

I've made https://pypi.python.org/pypi/signalfd since then (BSD).

**Comment by auvipy** [comment]
considering this an an extension

**Comment by ionelmc** [comment]
@auvipy do you have a more concrete plan regarding this alternate pool implementation?

**Comment by auvipy** [comment]
I have a broader plan. need some time to come with a CEP with that

**Comment by thedrow** [comment]
Thank you for your contribution. I'm sorry we couldn't proceed with merging it.

**Pull request closed without merging**